110 js form transition

billythakidd04 opened the pull request:
This should now work. If a user is unknown (not in the db) when they try to enter a review, the new user form should display.

In testing:
- create a review with a known (in db) user and make sure the review is created and attached to the correct user
- create a review with an unknown (**not** in db) user and make sure the user is created, the review is created, and the review is attached to the correct user

margeschrec committed:
js api forms rework, for #110

billythakidd04 committed:
for #110, this is almost finished, we need to get the usthe user id before submitting a review and its done

billythakidd04 committed:
for #110, remove ajax:false

billythakidd04 committed:
for #110, added new review as a callback function to newuser in js

margeschrec commented on js/reviewApi.js:
```diff
@@ -1,27 +1,143 @@
+	// check user route by email to see if the user exists
+	$.ajax(getUserSettings)
+		.done(function (response) { // successful response
+			console.log("success", response);
```
I can see in the console.log that the call is able to confirm and see that the user exists when the review form is given an email already in the db, however it then prompts the user to the create user form again.

billythakidd04 replied:
@margeschrec do you have some time this evening we can go over this and try to debug it?

margeschrec replied:
yes, I have time

margeschrec reviewed (changes requested):
forms update the db as intended however I can use the same email and it doesn't recognize that a user already exists by that email. Prompting me to create a new user.

billythakidd04 committed:
for #110, fixed new user form issue with $.when() call

billythakidd04 committed:
for #110 minor updates to processing. needs cleanup later

margeschrec commented:
Ok, went to check this again this morning and I'm getting a 404 not found when I enter a new email.  consolelog network  response message "user not found", it does not change the form or save the review to the db.

billythakidd04 commented:
> Ok, went to check this again this morning and I'm getting a 404 not found when I enter a new email. consolelog network response message "user not found", it does not change the form or save the review to the db.

@margeschrec a 404 is expected on the first attempt. But you're saying it doesn't save the user or review at all? Which form isn't showing. I need more info. Can you lay out your steps and possibly some screen shots or something to better explain your issue?

margeschrec commented:
> > Ok, went to check this again this morning and I'm getting a 404 not found when I enter a new email. consolelog network response message "user not found", it does not change the form or save the review to the db.
> 
> @margeschrec a 404 is expected on the first attempt. But you're saying it doesn't save the user or review at all? Which form isn't showing. I need more info. Can you lay out your steps and possibly some screen shots or something to better explain your issue?


![image](https://user-images.githubusercontent.com/78566744/115166620-0da45c00-a082-11eb-938e-3a21413e831d.png)

So if I start a new review with a new email like here with 4@4.4, after I click submit I get the "please sign up to submit the review" alert window at the top of the page and the form does not change to the create new user form, it dose save what was written in the review form...

![image](https://user-images.githubusercontent.com/78566744/115166748-b783e880-a082-11eb-8044-20305525ecb5.png)

and its interesting because now I no longer get the 404 error in the console. 

https://www.awesomescreenshot.com/video/3428119?key=4884d3bd1462a0b106f5e8f38b36d9f0

When I tested with an email already in the db it creates a new review as planned (without creating a new user with the same email- which is what we want).

vricci518 commented:
Ran a test review using an existing user already in the database. The review does not get accepted, the form does not clear.
When entering an email that does not exist in the database it does the same thing. Does not give alert window to create a new user account.

billythakidd04 commented:
yea Im testing now... Then def worked when I pushed up my last set of changes :roll_eyes:

billythakidd04 committed:
for #110, had to switch to $.ajax.then(successCallback(),errorCallback()) now that we properly return 404 errors for user/email route

billythakidd04 commented:
ok this should be fixed now. Since we also changed how we handle the errors for not found users (ie return a proper 404) the `$.done()` function doesn't handle the error. Switched to `$.then(successCallback, failureCallback)` and refactored the functions accordingly.

https://api.jquery.com/jquery.when/
https://api.jquery.com/deferred.then/#deferred-then-doneCallbacks-failCallbacks

margeschrec reviewed (dismissed):
Looking good now!

vricci518 reviewed (changes requested):
When I entered a review with an existing user I got the error message below and it went to the create user form.
![Screen Shot 2021-04-19 at 6 28 16 PM](https://user-images.githubusercontent.com/78501210/115311531-c9769180-a13d-11eb-8c06-4555fa0e0d88.png)
When I entered a review with as new user i got a message that said broke it with the following error message.
![Screen Shot 2021-04-19 at 6 31 51 PM (2)](https://user-images.githubusercontent.com/78501210/115311614-ef039b00-a13d-11eb-99d0-bc101edef80b.png)

margeschrec committed:
Merge branch 'master' into 110-JS-form-transition

nburd90 commented on forms/review.form.php:
```diff
@@ -44,46 +44,7 @@
 		// if userEmail is empty, add error message to array to be displayed later
 		$error['userEmail'] = 'Please Enter you Email';
 	}
```
@billythakidd04 these are the changes I pulled into my branch. https://github.com/eiman-kased/starbulls/pull/123#discussion_r617160183

billythakidd04 replied:
interesting... from where?

billythakidd04 replied:
I see... I'm not sure why this change was made. Let me review this pr really quick.

billythakidd04 replied:
Since I created this pr i cant request changes. can you please note that this needs to be fixed and return the php in the form?

billythakidd04 replied:
I dont want that change mistakenly slipping through and having to fix it later.

billythakidd04 commented:
@margeschrec I had not noticed until looking into #113 but the initial commit to this branch is removing all of the backend validation. That needs to be undone. If the user turns off js the php is our only protection and manner of submitting.

nburd90 reviewed (dismissed):
Return php validation in form per @billythakidd04

margeschrec committed:
re-added php backend validation to review.form.php

margeschrec commented:
> @margeschrec I had not noticed until looking into #113 but the initial commit to this branch is removing all of the backend validation. That needs to be undone. If the user turns off js the php is our only protection and manner of submitting.

Yes, big oversight on my part- fixing this now

margeschrec committed:
Merge branch 'master' into 110-JS-form-transition

margeschrec committed:
Merge branch 'master' into 110-JS-form-transition

eiman-kased reviewed (changes requested):
We need a way to tell the user that their review was successfully submitted. Maybe hide the form and show a Thank You for your review message.

margeschrec committed:
Merge branch 'master' into 110-JS-form-transition

margeschrec committed:
js review debug

margeschrec committed:
for #110 ready to be tested

billythakidd04 commented:
> We need a way to tell the user that their review was successfully submitted. Maybe hide the form and show a Thank You for your review message.

@eiman-kased this should be implemented. Please rereview.

eiman-kased commented:
> > We need a way to tell the user that their review was successfully submitted. Maybe hide the form and show a Thank You for your review message.
> 
> @eiman-kased this should be implemented. Please rereview.

This works!

eiman-kased reviewed (approved):
It works!

billythakidd04 commented:
i approve rubber stamping this as we tested as a group
